### ISSUE TITLE
GraphQL Introspection Queries 

### ISSUE THREAD

**Issue by csfreak92** [OWASP/ASVS#928]
I was thinking of a proposal to add a requirement in ASVS to ensure GraphQL introspection queries are disabled in production environment. We could argue that 14.3.2 - ` Verify that web or application server and application framework debug modes 
are disabled in production to eliminate debug features, developer consoles, 
and unintended security disclosures.` should cover it, however introspection queries are not just a debug feature and they could do more harm as they are a schema of the back-end API which helps any attacker map the API, particular parameters, etc. A debug mode such as stack traces do not give that much detail compared to introspection queries in my opinion. 

Also, I would like to raise awareness about introspection queries in GraphQL as many developers and DevOps people are not aware that this could be used against them by attackers. I do believe we should call this out as a separate security requirement/issue and not part of the generic 14.3.2. 

If we are in agreement, I would add it as `13.4.3` - `Verify that GraphQL introspection queries are disabled in production environment unless it is a public API where it is being used for integrations to your system`.

**Comment by Sjord** [comment]
> your system

I think it's unclear to who "your" refers here, or what "your system" is.

> unless it is a public API

API's that are meant for partners or third-party suppliers can also benifit from introspection queries, but are not necessarily public.

Perhaps we can rephrase this as "unless the GraphQL API is meant to be used by other clients"?

**Comment by csfreak92** [comment]
@Sjord, good catch. I was struggling to make it better but this sounds better. Though may I add, should we use the term "other parties" instead of "clients"? Clients could be mistaken for browsers or anything else.

**Comment by elarlang** [comment]
"your" is not good point of view - it's like you ask it from your friend who is developer, but requirement should be neutral and pragmatic. "You" is used only in one requirement and "your" is not used in any requirement.

**Comment by csfreak92** [comment]
I agree. I think it should be worded this way, `Verify that GraphQL introspection queries are disabled in production environment unless the GraphQL API is meant to be used by other parties`. I use parties instead of clients to prevent mistaking it from browsers or any client-side technology.
